### PR TITLE
💄(glimpse) add template for organization plugin

### PR DESCRIPTION
**Pull request by 2018D**
## Purpose

We want to allow showing a difference between founding members and new partners by changing the logo size.

## Proposal

Add a ~template~ variant option to differentiate between various types of organizations.

### CODE REVIEW

File: src/richie/apps/courses/migrations/0006_auto_20190807_2043.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    dependencies = [("courses", "0005_auto_20190717_0827")]
+
+    operations = [
+        migrations.AlterUniqueTogether(
+            name="persontitletranslation", unique_together=set()
+        ),
+        migrations.RemoveField(model_name="persontitletranslation", name="master"),
+        migrations.RemoveField(model_name="person", name="first_name"),
+        migrations.RemoveField(model_name="person", name="last_name"),
+        migrations.RemoveField(model_name="person", name="person_title"),
```
Comment (sampaccoud) [inline src/richie/apps/courses/migrations/0006_auto_20190807_2043.py]: @2018D these modifications are not related to your PR so don't include them.
We are planning to remove these fields in the next BC release.

File: src/richie/apps/courses/migrations/0006_auto_20190807_2043.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                    ("courses/plugins/organization_small.html", "Small"),
+                ],
+                default="courses/plugins/organization.html",
+                help_text="Optional template for custom look.",
+                max_length=150,
+                verbose_name="Template",
+            ),
+        ),
+        migrations.DeleteModel(name="PersonTitle"),
+        migrations.DeleteModel(name="PersonTitleTranslation"),
```
Comment (sampaccoud) [inline src/richie/apps/courses/migrations/0006_auto_20190807_2043.py]: ditto

File: src/richie/apps/courses/cms_plugins.py [hunk 3]
```diff
@@ -27,9 +28,10 @@ class OrganizationPlugin(CMSPluginBase):
     """
 
     cache = True
+    fieldsets = ((None, {"fields": ["page", "template"]}),)
     model = OrganizationPluginModel
     module = PLUGINS_GROUP
-    render_template = "courses/plugins/organization.html"
+    render_template = ORGANIZATION_TEMPLATES[0][0]
```
Comment (sampaccoud) [inline src/richie/apps/courses/cms_plugins.py]: maybe this is not necessary since you are overriding it below?

File: src/richie/apps/courses/templates/courses/cms/fragment_organization_small_glimpse.html [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{% load i18n cms_tags extra_tags thumbnail %}
```
Comment (sampaccoud) [inline src/richie/apps/courses/templates/courses/cms/fragment_organization_small_glimpse.html]: since this is a massive copy/paster, have you tried factorizing it and forcing a context variable like it's done for categories (tag, icon...)?

File: src/richie/apps/courses/cms_plugins.py (hunk 3 above)
Comment (2018D) [reply src/richie/apps/courses/cms_plugins.py]: It's to force a default and make sure that previously added plugins still work.

File: src/richie/apps/courses/templates/courses/cms/fragment_organization_small_glimpse.html (hunk 4 above)
Comment (2018D) [reply src/richie/apps/courses/templates/courses/cms/fragment_organization_small_glimpse.html]: i have not. I'll look into it.

File: src/richie/apps/courses/cms_plugins.py (hunk 3 above)
Comment (sampaccoud) [reply src/richie/apps/courses/cms_plugins.py]: But you're adding a field to the model with a default... and anyway you always override it so this should be useless.

File: src/richie/apps/courses/templates/courses/cms/fragment_organization_small_glimpse.html (hunk 4 above)
Comment (2018D) [reply src/richie/apps/courses/templates/courses/cms/fragment_organization_small_glimpse.html]: I still need to call a different template because it's optional and not based on context.
I now use the fragment_organization more efficiently.

File: src/richie/apps/courses/cms_plugins.py (hunk 3 above)
Comment (2018D) [reply src/richie/apps/courses/cms_plugins.py]: It's required for the build. Copied a comment from section_plugin indicating it.

**Comment by 2018D** [comment]
Added a test to verify if the templates are rendered correctly.

File: src/richie/apps/courses/cms_plugins.py [hunk 5]
```diff
@@ -27,9 +28,12 @@ class OrganizationPlugin(CMSPluginBase):
     """
 
     cache = True
+    fieldsets = ((None, {"fields": ["page", "template"]}),)
     model = OrganizationPluginModel
     module = PLUGINS_GROUP
-    render_template = "courses/plugins/organization.html"
+    # Required from CMSPluginBase signature but not used since we override it
+    # from render()
```
Comment (sampaccoud) [inline src/richie/apps/courses/cms_plugins.py]: :+1:

File: src/richie/apps/courses/templates/courses/plugins/organization_small.html [hunk 6]
```diff
@@ -0,0 +1,10 @@
+{% load cms_tags i18n %}
+{% spaceless %}
+{% if current_page.publisher_is_draft or instance.check_publication is True %}
+  {% with organization_page=relevant_page.organization.extended_object %}
+    <a class="organization-glimpse__small organization-glimpse organization-glimpse--link{% if organization_page.publisher_is_draft is True %} organization-glimpse--draft{% endif %}" href="{{ organization_page.get_absolute_url }}">
```
Comment (mbenadda) [inline src/richie/apps/courses/templates/courses/plugins/organization_small.html]: The link is part of the template (it's inside the "component boundary") and should not be taken out. I previously fixed this because we had links sitting outside the component boundary.

If you want to apply a different class on some use sites it should be done through a prop you pass to the glimpse template

Comment (sampaccoud) [reply src/richie/apps/courses/templates/courses/plugins/organization_small.html]: @2018D good catch, and we already do it for the category here: https://github.com/openfun/richie/blob/master/src/richie/apps/courses/templates/courses/plugins/category_plugin.html

You should use the same pattern and pass a "form_factor" prop.

**Comment by sampaccoud** [comment]
@2018D I merged PR https://github.com/openfun/richie/pull/817 which added a [migration 0006](https://github.com/openfun/richie/blob/master/src/richie/apps/courses/migrations/0006_add_program.py) in the `courses` app.

Remember to rename your migration to 0007 when rebasing on master.

**Comment by sampaccoud** [comment]
Following discussions in this PR, I have refactored it to respect the glimpse. Instead of overriding the template, we should allow forcing a form factor via the context.

@mbenadda can you see if it now fits your requirements?
